modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistry.java

```diff
@@ -0,0 +1,196 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.metric;
+
+import java.util.function.BooleanSupplier;
+import java.util.function.DoubleSupplier;
+import java.util.function.IntSupplier;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.apache.ignite.internal.processors.metric.impl.DoubleMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HitRateMetric;
+import org.apache.ignite.internal.processors.metric.impl.IntMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongAdderMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.LongMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.BooleanMetricImpl;
+import org.apache.ignite.internal.processors.metric.impl.HistogramMetric;
+import org.apache.ignite.internal.processors.metric.impl.ObjectMetricImpl;
+import org.apache.ignite.spi.metric.BooleanMetric;
+import org.apache.ignite.spi.metric.IntMetric;
+import org.apache.ignite.spi.metric.LongMetric;
+import org.apache.ignite.spi.metric.Metric;
+import org.apache.ignite.spi.metric.ReadOnlyMetricRegistry;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Metrics registry.
+ * Provide methods to register required metrics, gauges for Ignite internals.
+ * Provide the way to obtain all registered metrics for exporters.
+ */
+public interface MetricRegistry extends ReadOnlyMetricRegistry {
+    /**
+     * Register existing metrics in this group with the specified name.
+     *
+     * @param metric Metric.
+     */
+    public void register(Metric metric);
+
+    /**
+     * Removes metrics with the {@code name} from registry.
+     *
+     * @param name Metric name.
+     */
+    public void remove(String name);
+
+    /**
+     * Resets state of this metric set.
+     */
+    public default void reset() {
+        getMetrics().forEach(Metric::reset);
+    }
+
+    /**
+     * Register {@link BooleanMetric} that value will be queried from specified supplier.
+     *
+     * @param name Name.
+     * @param supplier Supplier.
+     * @param description Description.
+     */
+    public void register(String name, BooleanSupplier supplier, @Nullable String description);
```

Comment:
What's the point in having register* set of methods ?
It can be replaced by obtaining writeable metric and setting value provided by closure on it.
Please provide more detailed documentation for use cases of this functionality.

Reply:
1. It is useful for monitoring some third-party entities.
`ExecutorService` is one of the examples.
Please, take a look into `GridMetricManager#monitorExecutor`.

2. Several existing metrics calculated at the moment they queried. You can find examples in `DataRegionMetricsImpl`: "TotalUsedPages", "TotalAllocatedSize". I think we can rewrite most of it by the way you mentioned, but this is the scope of the following tickets.